play-json/shared/src/main/scala-2.13+/play/api/libs/json/ImmutableLinkedHashMap.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2009-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.api.libs.json
+
+import java.util.{ LinkedHashMap => JLinkedHashMap }
+import scala.collection.AbstractIterator
+import scala.collection.MapFactory
+import scala.collection.immutable.AbstractMap
+import scala.collection.mutable
+
+/**
+ * Wraps a Java LinkedHashMap as a Scala immutable.Map.
+ */
+private[json] class ImmutableLinkedHashMap[A, +B](underlying: JLinkedHashMap[A, B]) extends AbstractMap[A, B] {
```

Comment:
Maybe this should be called `UnmodifiableLinkedHashMap` since it doesn't guarantee that other code can't change the underlying collection.

Reply:
Subtle distinction. The java `Unmodifiable` classes allow users to wrap mutable collections they control whereas the immutable ones don't (with the exception of `scala.collection.immtuable.ArraySeq.unsafeWrapArray`, `akka.util.ByteString.fromArrayUnsafe`).

This `ImmutableLinkedHashMap` package-protects the constructor (and the companion object). This seems at least as strong as the others. Technically, the user can use reflection to break any guarantees if they really want to.

I think it all comes down to intent. I read "unmodifiable" as "you can't modify it, but I can and might", and immutable as "I'm not gonna modify it either".

Reply:
```scala
val arr = Array[AnyRef]("Vector is immutable, right?")
val as = ArraySeq.unsafeWrapArray(arr)
val v = Vector.from(as)
println(v) // Vector(Vector is immutable, right?)
arr(0) = "This makes me uncomfortable"
println(v) // Vector(This makes me uncomfortable)
```

Reply:
yeah, it's an internal class so I'm not super concerned about it either way 🤷